### ISSUE TITLE
Feature: add RESTIC_DRY_RUN environment variable

### ISSUE THREAD

**Issue by codemicro** [restic/restic#3703]
Output of `restic version`
--------------------------

`restic 0.13.1 (v0.13.1-0-g594f155e) compiled with go1.18 on linux/amd64`

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

It might be helpful to have an environment variable which enables/disables the dry-run mode, similar to the `RESTIC_REPOSITORY` or `RESTIC_PASSWORD` variables, to complement the `--dry-run`/`-n` flag.

What are you trying to do? What problem would this solve?
---------------------------------------------------------

This would be useful, for example, in a multi-stage Bash script where Restic is invoked multiple times. Instead of adding/removing the `--dry-run` flag in multiple places, one could just add/remove `export RESTIC_DRY_RUN=true` or similar.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Restic is awesome, and has been helping me get my crap in order and hopefully preventing me from losing my photo archive in the future! :)

**Comment by MichaelEischer** [comment]
I see two problems with that:
- not all command implement a dry-run mode. Now, if you specify the `--dry-run` option for a command that doesn't support it, this will result in an error. With an environment variable, the variable would just be ignored and restic would unexpectedly modify the repository.
- With an environment variable it would be much harder to see whether a command runs in dry-run mode. But for most dry-run use cases, I really want to be sure that the mode is active.

**Comment by codemicro** [comment]
* > Now, if you specify the --dry-run option for a command that doesn't support it, this will result in an error

  Granted I don't know anything about the internals of Restic and how it works, but couldn't the environment variable being set cause the same reaction? Is it possible to have a command fail in the same way as it would if `--dry-run` was specified if `RESTIC_DRY_RUN` specified?

* > I really want to be sure that the mode is active.

  Could a big warning as soon as the command starts be useful for this? `WARNING: Restic is running in dry run mode because RESTIC_DRY_RUN is set. No changes will be made to the repository, etc, etc, etc` Colours could even be added to it to give it some extra emphasis.

**Comment by ethereal-engineer** [comment]
It could be useful to extend the dry run feature to any action that would cause data change. I am (and have been for some months now) writing a centrally-controlled and monitored secure site backup solution based on restic, and the ability to test using dry run on all potentially modifying commands would be a plus.

**Comment by codemicro** [comment]
While I agree, that's outside the scope of this issue.

**Comment by ethereal-engineer** [comment]
True, but it lends credence to the idea of a globally-scoped RESTIC_DRY_RUN flag 😉

**Comment by rawtaz** [comment]
Let's all take a step back here :) @codemicro What is your *actual* use case? What are you trying to or need to do?

I read what you wrote initially and I don't really see the/a need for this flag in that description:

> This would be useful, for example, in a multi-stage Bash script where restic is invoked multiple times. Instead of adding/removing the `--dry-run` flag in multiple places, one could just add/remove `export RESTIC_DRY_RUN=true` (or at least, I think that's right).

You can simply make a variable in your script, e.g. `BACKUP_ARGS=""` and make all the invocations to `restic backup` have `$BACKUP_ARGS` in the command line, e.g. `restic -r foo backup $BACKUP_ARGS /my/stuff`. When the variable has an empty value (which it would normally do) the final command that's invoked would be `restic -r foo backup  /my/stuff`, and if you modify the value of the variable in that script of yours to be `--dry-run`, then the final invoked command would be `restic -r foo backup --dry-run /my/stuff`.

In other words, there's no need for an actual environment variable to reach the goal of not having to modify multiple lines every time you want to turn on/off the dry run.

**Comment by codemicro** [comment]
The actual use case was testing a Bash script containing calls to `backup` and `prune`. Though, it hadn't crossed my mind that it's possible to do the variable interpolation like that. I'll most likely end up doing that, so thanks for the heads up!

That said, I'd still argue that it's worth having a `RESTIC_DRY_RUN` variable of some sort. It might not be strictly neccessary, but I think it'd be helpful (though I have had a quick look through the code and appreciate it's not as simple as it might otherwise sound to add).

**Comment by MichaelEischer** [comment]
> * Could a big warning as soon as the command starts be useful for this? `WARNING: Restic is running in dry run mode because RESTIC_DRY_RUN is set. No changes will be made to the repository, etc, etc, etc` Colours could even be added to it to give it some extra emphasis.

That direction is not the problem i was worried about. If restic accidentally runs in dry-run mode the worst outcome is wasting time (unless that prevents the backup from ever succeeding, but you want monitoring for that anyways). What I'm more worried about is that someone runs `RESTIC_DRY_RUN=true` in their terminal, starts messing around with e.g. `forget` and afterwards wonders why most snapshots have been delete. What went wrong? The environment variable was not `export`ed. We've seen quite a few bug reports with users messing that up...

**Comment by codemicro** [comment]
Yeah, that is an issue, and not one I'd considered.

**Comment by mgumz** [comment]
@MichaelEischer the argument about having BACKUP_ARGS is a valid one. 

the argument about people having set an environment variable by accident and now having it forgotten … is a bit shaky: RESTIC_REPOSITORY etc … why do they exist then in the first place if people could get confused about the environment variable and the "-r" flag for example.

i'll now switch to BACKUP_ARGS in my script and dump https://github.com/restic/restic/pull/4072
